DEVELOPER.md

```diff
@@ -209,6 +256,9 @@ In the `/src` directory you will find:
   - Crates here contain **specific implementations of services and repositories** (e.g. repository that stores data in S3)
   - Crate directories are named as `<domain>-<technology>`, e.g. `object-repository-s3` while crate names will typically have `kamu-<domain>-<technology>` prefix
   - Infrastructure layer only operates on entities and interfaces defined in `domain` layer
+- `database`
```

Comment:
> Migrations for an app would therefore be a union of migrations of all repos it depends on.

The existing scripting in `sqlx-cli` will not allow it, unless there is some process that temporarily merges all migration scripts into a single folder before applying. Besides, we might have much more complex data-level dependencies, such as a table referring to a foreign key from another domain. Even if domains logically have only a weak reference, at the database level, we would likely wan't to enforce some integrity constraints.

> Keeping all repos in separate crates would allow us to easily chose which systems should or shouldn't be deployed together on an app level. E.g. we can deploy them as an api-server monolith, or split them into microservices.

I understand the motivation. The complication is that every database-specific crate holding repository implementation would require maintaining a dedicated `.env` file pointing to correct `DATABASE_URL`. If they would be scattered across the project folders, than we would need makefiles to be able to generate all of them in 1 command.

> But perhaps we could have them in a single utils/db-common crate and enable pg and other specific wrappers as cargo features?

Must be doable, assuming this `db-common` would not have any queries, so that we don't need to supply a single `DATABASE_URL` for this crate. Without queries, concentrating all database-specific types in 1 crate should be possible we would just enable more features in `sqlx` dependency.

>  E.g. we can deploy them as an api-server monolith, or split them into microservices.

Microservices normally use separate databases. It's even considered an anti-pattern, if multiple services write to the same database. Thinking in microservices is a radical architectural change, we should be aware of earlier than sooner. The impact on persistence design is too serious.

Reply:
One thing you did cover here are data models. Right now they are in the wrong place in `database-common`, but they are also domain-specific, and have to be placed somewhere.

Reply:
> The complication is that every database-specific crate holding repository implementation would require maintaining a dedicated .env file pointing to correct DATABASE_URL. If they would be scattered across the project folders, than we would need makefiles to be able to generate all of them in 1 command.

Yep, this is pretty much what I just suggested on my other comment about offline mode - generating and cleaning up `.env` per repository crate in makefiles.

> Microservices normally use separate databases

Yes, that's what I meant. Repo crates would define their own few tables and they are either created in a database of a monolith, or, for example, we take the entire flow system and deploy it as a micro-service with the same tables but in its own database.

> temporarily merges all migration scripts into a single folder before applying

Migrations are weird. They seem more like a deployment artifact than code, and relate much closer to an application than to the crates that it's composed of. Perhaps repository crates should only provide one migration - one that initializes DB for testing in its final state. And "real" migrations that manage the **evolution** of the database will be on an application level - where they will be tied to app's release cycle and to how it actually gets deployed.

Looking at how migrations are versioned - an approach with a script that copies them from individual crates into `app/migrations` folder might work too. Perhaps it can even co-exist with app-specific migrations: e.g. all repo crates create tables that are fully independent, but there's an app-specific migration that later adds ForeignKey constraints between them to ensure some inter-aggregate consistency.